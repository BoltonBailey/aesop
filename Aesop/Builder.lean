/-
Copyright <redacted>
Released under Apache 2.0 license as described in the file LICENSE.
Authors: <redacted>
-/

import Aesop.Builder.Apply
import Aesop.Builder.Basic
import Aesop.Builder.Cases
import Aesop.Builder.Constructors
import Aesop.Builder.Default
import Aesop.Builder.Forward
import Aesop.Builder.NormSimp
import Aesop.Builder.Tactic
