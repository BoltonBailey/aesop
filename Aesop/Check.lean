/-
Copyright (c) 2021 Jannis Limperg. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Jannis Limperg
-/

import Lean

open Lean

namespace Aesop

def registerCheckOption [KVMap.Value α] (checkName : Name) (defValue : α)
    (descr : String) : IO (Lean.Option α) :=
  Option.register (`aesop.check ++ checkName)
    { defValue := defValue, group := "aesop", descr := descr }

initialize checkAllOption : Lean.Option Bool ←
  registerCheckOption `all false
    "(aesop) Enable all runtime checks. Individual checks can still be disabled."

initialize checkProofReconstructionOption : Lean.Option Bool ←
  registerCheckOption `proofReconstruction false
    "(aesop) Typecheck partial proof terms during proof reconstruction."

inductive Check
  | all
  | proofReconstruction

namespace Check

def toOption : Check → Lean.Option Bool
  | all => checkAllOption
  | proofReconstruction => checkProofReconstructionOption

def isEnabled [Monad m] [MonadOptions m] : Check → m Bool
  | all => return all.toOption.get (← getOptions)
  | c => do
    let opts ← getOptions
    match c.toOption.get? opts with
    | none => return all.toOption.get opts
    | some v => return v

end Check

end Aesop
