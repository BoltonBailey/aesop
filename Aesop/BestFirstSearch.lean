/-
Copyright (c) 2021 Jannis Limperg. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Jannis Limperg, Asta Halkjær From
-/

import Aesop.Rule
import Aesop.Tree
import Aesop.Util

open Lean
open Lean.Meta
open Std (BinomialHeap)

namespace Aesop.Search

structure ActiveGoal where
  goal : GoalRef
  successProbability : Percent

namespace ActiveGoal

instance : LT ActiveGoal where
  lt g h := g.successProbability < h.successProbability

instance : DecidableRel (α := ActiveGoal) (· < ·) :=
  λ g h =>
    inferInstanceAs (Decidable (g.successProbability < h.successProbability))

def ofGoalRef [Monad m] [MonadLiftT (ST IO.RealWorld) m] (gref : GoalRef) :
    m ActiveGoal :=
  return {
    goal := gref
    successProbability := (← gref.get).successProbability
  }

end ActiveGoal

abbrev ActiveGoalQueue := BinomialHeap ActiveGoal (· > ·)
  -- ActiveGoals are ordered by success probability. Here we want highest
  -- success probability first.

structure Context where
  ruleSet : RuleSet
  mainGoal : MVarId

structure State where
  tree : Tree
  activeGoals : ActiveGoalQueue

namespace State

def mkInitial (t : Tree) : State where
  tree := t
  activeGoals := BinomialHeap.empty

def mkInitialForMainGoal (mainGoal : MVarId) : MetaM State := do
  let goal ← copyMVar mainGoal
  let tree ← Tree.singleton $
    Goal.mk none #[] $ GoalData.mkInitial goal Percent.hundred
  return State.mkInitial tree

end State

abbrev SearchM := ReaderT Context $ StateRefT State MetaM

namespace SearchM

def run (ctx : Context) (state : State) (x : SearchM α) : MetaM (α × State) :=
  StateRefT'.run (ReaderT.run x ctx) state

end SearchM

instance (priority := 0) : MonadReaderOf RuleSet SearchM where
  read := Context.ruleSet <$> read

instance (priority := 0) : MonadStateOf Tree SearchM :=
  MonadStateOf.ofLens State.tree (λ t s => { s with tree := t })

instance (priority := 0) : MonadStateOf ActiveGoalQueue SearchM :=
  MonadStateOf.ofLens State.activeGoals (λ an s => { s with activeGoals := an })

def readMainGoal : SearchM MVarId :=
  Context.mainGoal <$> read

def addGoal (goal : MVarId) (successProbability : Percent)
    (parent : RappRef) : SearchM GoalRef := do
  let g ← GoalData.mkInitial goal successProbability
  let t ← getThe Tree
  let (_, gref, t) ← t.insertGoal g parent
  setThe Tree t
  modifyThe ActiveGoalQueue λ q => q.insert
    { goal := gref
      successProbability := successProbability }
  return gref

-- TODO array?
def addGoals (goals : List MVarId) (successProbability : Percent)
    (parent : RappRef) : SearchM (Array GoalRef) := do
  let mut acc := #[]
  for goal in goals do
    let gref ← addGoal goal successProbability parent
    acc := acc.push gref
  return acc

def runNormalizationRule (goal : MVarId) (r : NormalizationRule) :
    SearchM MVarId := do
  let subgoals ← try r.tac goal catch e => throwError
    "aesop: normalization rule {r.name} failed with error:\n{e.toMessageData}"
    -- TODO show error context
  match subgoals with
  | [g] => return g
  | _ => throwError "aesop: normalization rule {r.name} did not produce exactly one subgoal"

def runNormalizationSimp (goal : MVarId) (ctx : Simp.Context) : SearchM MVarId := do
  let (some goal) ← simpAll goal ctx | throwError
    "aesop: normalization simp rule solved the goal"
  return goal

def normalizeGoalMVar (goal : MVarId) : SearchM MVarId := do
  let rs ← readThe RuleSet
  let rules ← rs.applicableNormalizationRules goal
  let (preSimpRules, postSimpRules) :=
    rules.partition λ r => r.priorityInfo < (0 : Int)
  let goal ← preSimpRules.foldlM (init := goal) runNormalizationRule
  let simpCtx :=
    { (← Simp.Context.mkDefault) with simpLemmas := rs.normalizationSimpLemmas }
  let goal ← runNormalizationSimp goal simpCtx
  let goal ← postSimpRules.foldlM (init := goal) runNormalizationRule
  return goal

def normalizeGoalIfNecessary (gref : GoalRef) : SearchM Unit :=
  gref.modifyM λ g => do
    let (false) ← pure g.normal? | return g
    let newGoal ← normalizeGoalMVar g.goal
    return g.setGoal newGoal

def runRule (goal : MVarId) (r : MVarId → MetaM (List MVarId)) :
    MetaM (Option (MVarId × List MVarId)) := do
  let proofMVar ← copyMVar goal
  let subgoals ← observing? $ r proofMVar
  return subgoals.map (proofMVar, ·)

inductive RuleResult
| proven
| failed
| succeeded

namespace RuleResult

def failed? : RuleResult → Bool
  | failed => true
  | _ => false

end RuleResult

def applyRegularRule (parentRef : GoalRef) (rule : RegularRule) :
    SearchM RuleResult := do
  let state ← getThe State
  let parent ← parentRef.get
  let result ← runRule parent.goal rule.tac
  let successProbability :=
    parent.successProbability * rule.successProbability
  match result with
  | some (proofMVar, []) => do
    -- Rule succeeded and did not generate subgoals, meaning the parent
    -- node is proven.
    let r :=
      { RappData.mkInitial rule successProbability (mkMVar proofMVar) with
        proven? := true }
    let (_, _, tree) ← state.tree.insertRapp r parentRef
    setThe Tree tree
    parentRef.setProven
    return RuleResult.proven
  | some (proofMVar, subgoals) => do
    -- Rule succeeded and generated subgoals.
    let r := RappData.mkInitial rule successProbability (mkMVar proofMVar)
    let (_, rappRef, tree) ← state.tree.insertRapp r parentRef
    setThe Tree tree
    let _ ← addGoals subgoals successProbability rappRef
    return RuleResult.succeeded
  | none => do
    -- Rule did not succeed.
    parentRef.modify λ (g : Goal) => g.setFailedRapps $ rule :: g.failedRapps
    parentRef.setUnprovable
    return RuleResult.failed

def applyFirstSafeRule (gref : GoalRef) : SearchM RuleResult := do
  let g ← gref.get
  let rules ← (← readThe RuleSet).applicableSafeRules g.goal
  let mut result := RuleResult.failed
  for r in rules do
    result ← applyRegularRule gref $ RegularRule.safe r
    if result.failed? then continue else break
  return result

def selectRules (gref : GoalRef) : SearchM (List UnsafeRule) := do
  let g ← gref.get
  match g.unsafeQueue with
  | some rules => return rules
  | none => do
    let rs ← readThe RuleSet
    let rules := (← rs.applicableUnsafeRules g.goal).toList
    gref.set $ g.setUnsafeQueue rules
    return rules

def applyFirstUnsafeRule (gref : GoalRef) : SearchM Bool := do
  let rules ← selectRules gref
  let mut result := RuleResult.failed
  let mut remainingRules := rules
  for r in rules do
    remainingRules := remainingRules.tail!
    result ← applyRegularRule gref (RegularRule.unsafe r)
    if result.failed? then continue else break
  gref.modify λ g => g.setUnsafeQueue remainingRules
  if result.failed? && remainingRules.isEmpty then gref.setUnprovable
  return ¬ remainingRules.isEmpty

def expandGoal (gref : GoalRef) : SearchM Bool := do
  normalizeGoalIfNecessary gref
  let result ← applyFirstSafeRule gref
  if result.failed?
    then applyFirstUnsafeRule gref
    else pure false

def expandNextGoal : SearchM Unit := do
  let some (activeGoal, activeGoals) ← pure (← getThe ActiveGoalQueue).removeMin
    | throwError "aesop/expandNextGoal: internal error: no active goals left"
  setThe ActiveGoalQueue activeGoals
  let gref := activeGoal.goal
  let g ← gref.get
  unless g.proven? ∨ g.unprovable? ∨ g.irrelevant? do
    let hasMoreRules ← expandGoal gref
    if hasMoreRules then do
      let ag ← ActiveGoal.ofGoalRef gref
      modifyThe ActiveGoalQueue λ q => q.insert ag

def finishIfProven : SearchM Bool := do
  let t ← getThe Tree
  let (true) ← pure (← t.root.get).proven?
    | return false
  t.linkProofs
  let prf ← t.extractProof
  assignExprMVar (← readMainGoal) prf
  return true

partial def search : SearchM Unit := do
  let t ← getThe Tree
  let (false) ← t.rootUnprovable?
    | throwError "aesop: failed to prove the goal"
  let done ← finishIfProven
  if ¬ done then
    expandNextGoal
    search

end Search

def search (rs : RuleSet) (mainGoal : MVarId) : MetaM Unit := do
  let state ← Search.State.mkInitialForMainGoal mainGoal
  let ctx := { ruleSet := rs, mainGoal := mainGoal }
  let _ ← Search.search.run ctx state

end Aesop
