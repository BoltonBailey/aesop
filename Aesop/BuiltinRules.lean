/-
Copyright (c) 2021 Jannis Limperg. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Jannis Limperg
-/

-- The Aesop.BuiltinRules.* imports are needed to ensure that the tactics from
-- these files are registered.
import Aesop.BuiltinRules.Assumption
import Aesop.BuiltinRules.ApplyHyps
import Aesop.BuiltinRules.Reflexivity
import Aesop.BuiltinRules.SplitHyps
import Aesop.Frontend

open Lean
open Lean.Elab
open Lean.Elab.Tactic

namespace Aesop.BuiltinRules

-- TODO avoid TacticM?
@[aesop norm -1 (tactic (uses_branch_state := false)) (rule_sets [builtin])]
def intros : TacticM Unit := do
  evalTactic (‚Üê `(tactic|intros))

end Aesop.BuiltinRules
