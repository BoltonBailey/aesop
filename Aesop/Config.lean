/-
Copyright (c) 2021 Jannis Limperg. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Jannis Limperg, Asta Halkjær From
-/

import Aesop.RuleSet
import Aesop.RuleBuilder
import Aesop.Options

open Lean
open Lean.Meta
open Lean.Elab.Term

namespace Aesop

namespace Parser.Attribute

declare_syntax_cat aesop_prio

syntax "-"? num "%"? : aesop_prio

declare_syntax_cat' aesop_kind force_leading_unreserved_tokens

syntax (aesop_prio)? : aesop_kind
syntax &"safe" (aesop_prio)? : aesop_kind
syntax &"unsafe" (aesop_prio)? : aesop_kind
syntax &"norm" (aesop_prio)? : aesop_kind

declare_syntax_cat' aesop_builder force_leading_unreserved_tokens

syntax &"apply" : aesop_builder
syntax &"simp" : aesop_builder
syntax &"unfold" : aesop_builder
syntax &"tactic" : aesop_builder
syntax &"constructors" : aesop_builder
syntax &"safe_default" : aesop_builder
syntax &"unsafe_default" : aesop_builder
syntax &"norm_default" : aesop_builder

declare_syntax_cat' aesop_clause

syntax "(" &"builder" aesop_builder ")" : aesop_clause
syntax "(" &"options" term ")" : aesop_clause

syntax (name := aesop) &"aesop" aesop_kind aesop_clause* : attr

end Parser.Attribute


variable [Monad m] [MonadError m]


inductive Prio
  | successProbability (p : Percent)
  | penalty (i : Int)
  deriving Inhabited, BEq

namespace Prio

instance : ToString Prio where
  toString
    | successProbability p => p.toHumanString
    | penalty i => toString i

protected def parse : Syntax → Except String Prio
  | `(aesop_prio|- $n:numLit %) =>
    throw "percentage cannot be negative"
  | `(aesop_prio|- $n:numLit) => return penalty $ - n.toNat
  | `(aesop_prio|$n:numLit) => return penalty $ n.toNat
  | `(aesop_prio|$n:numLit %) => do
    let (some p) ← Percent.ofNat n.toNat
      | throw "percentage must be between 0 and 100."
    return successProbability p
  | _ => unreachable!

end Prio

def parsePrioForUnsafeRule : Option Syntax → Except String Percent
  | none => throw "unsafe rule must be given a success probability."
  | some p => do
    let (Prio.successProbability p) ← Prio.parse p | throw
      "unsafe rule must be given a success probability ('x%'), not an integer penalty"
    return p

def parsePrioForSafeRule : Option Syntax → Except String Int
  | none => return defaultSafePenalty
  | some p => do
    let (Prio.penalty p) ← Prio.parse p | throw
      "safe rule must be given an integer penalty, not a success probability"
    return p

def parsePrioForNormRule : Option Syntax → Except String Int
  | none => return defaultNormPenalty
  | some p => do
    let (Prio.penalty p) ← Prio.parse p | throw
      "norm rule must be given an integer penalty, not a success probability"
    return p

inductive RuleKind
  | norm (penalty : Int)
  | safe (penalty : Int)
  | «unsafe» (successProbability : Percent)
  deriving Inhabited, BEq

namespace RuleKind

instance : ToString RuleKind where
  toString
    | norm p => s!"norm {p}"
    | safe p => s!"safe {p}"
    | «unsafe» p => s!"unsafe {p.toHumanString}"

protected def parse : Syntax → m RuleKind
  | `(aesop_kind|safe $[$prio:aesop_prio]?) =>
    go (parsePrioForSafeRule prio) safe
  | `(aesop_kind|unsafe $[$prio:aesop_prio]?) =>
    go (parsePrioForUnsafeRule prio) «unsafe»
  | `(aesop_kind|$[$prio:aesop_prio]?) =>
    go (parsePrioForUnsafeRule prio) «unsafe»
  | `(aesop_kind|norm $[$prio:aesop_prio]?) =>
    go (parsePrioForNormRule prio) norm
  | _ => unreachable!
  where
    go {α} (prio : Except String α) (cont : α → RuleKind) : m RuleKind :=
      match prio with
      | Except.ok prio => return cont prio
      | Except.error e => throwError "aesop: {e}"

end RuleKind


inductive RegularBuilderClause
  | apply
  | tactic
  | constructors
  | unsafeDefault
  | safeDefault
  deriving Inhabited, BEq

namespace RegularBuilderClause

def builderName : RegularBuilderClause → String
  | apply => "apply"
  | tactic => "tactic"
  | constructors => "constructors"
  | unsafeDefault => "unsafe_default"
  | safeDefault => "safe_default"

instance : ToString RegularBuilderClause where
  toString c := s!"(builder {c.builderName})"

def toRuleBuilder : RegularBuilderClause → RuleBuilder RegularRuleBuilderResult
  | apply => RuleBuilder.apply
  | tactic => RuleBuilder.tactic
  | constructors => RuleBuilder.constructors
  | unsafeDefault => RuleBuilder.unsafeRuleDefault
  | safeDefault => RuleBuilder.safeRuleDefault

end RegularBuilderClause


inductive BuilderClause
  | regular (c : RegularBuilderClause)
  | simpLemma
  | simpUnfold
  | normDefault
  deriving Inhabited, BEq

namespace BuilderClause

def builderName : BuilderClause → String
  | regular c => c.builderName
  | simpLemma => "simp"
  | simpUnfold => "unfold"
  | normDefault => "norm_default"

instance : ToString BuilderClause where
  toString c := s!"(builder {c.builderName})"

open RegularBuilderClause in
protected def parseBuilder : Syntax → BuilderClause
  | `(aesop_builder|apply) => regular apply
  | `(aesop_builder|tactic) => regular tactic
  | `(aesop_builder|constructors) => regular constructors
  | `(aesop_builder|simp) => simpLemma
  | `(aesop_builder|unfold) => simpUnfold
  | `(aesop_builder|unsafe_default) => regular unsafeDefault
  | `(aesop_builder|safe_default) => regular safeDefault
  | `(aesop_builder|norm_default) => normDefault
  | _ => unreachable!

def toRuleBuilder : BuilderClause → RuleBuilder NormRuleBuilderResult
  | regular c => λ goal i =>
    NormRuleBuilderResult.regular <$> c.toRuleBuilder goal i
  | simpLemma => RuleBuilder.normSimpLemmas
  | simpUnfold => RuleBuilder.normSimpUnfold
  | normDefault => RuleBuilder.normRuleDefault

end BuilderClause


inductive Clause
  | builder (c : BuilderClause)
  deriving Inhabited, BEq

namespace Clause

instance : ToString Clause where
  toString
    | builder c => toString c

protected def parse : Syntax → m Clause
  | `(aesop_clause|(builder $b:aesop_builder)) =>
    builder <$> BuilderClause.parseBuilder b
  | _ => unreachable!

end Clause


structure Clauses where
  builder : Option BuilderClause := none

namespace Clauses

def add (cs : Clauses) : Clause → Clauses
  | Clause.builder b =>
    { cs with builder := cs.builder.mergeRightBiased (some b) }

def ofClauseArray (cs : Array Clause) : Clauses :=
  cs.foldl add {}

end Clauses


structure NormRuleClauses where
  builder : BuilderClause

namespace NormRuleClauses

def ofClauses (cs : Clauses) : NormRuleClauses where
  builder := cs.builder.getD BuilderClause.normDefault

end NormRuleClauses

open NormRuleBuilderResult in
def buildNormRule (penalty : Int) (clauses : NormRuleClauses) (i : RuleIdent) :
    MetaM (Array RuleSetMember) := do
  let results ← clauses.builder.toRuleBuilder i
  match results with
  | regular results =>
    results.mapM λ res => RuleSetMember'.normRule
      { name := `norm ++ res.builderName ++ i.ruleName
        indexingMode := res.indexingMode
        usesBranchState := res.mayUseBranchState
        extra := { penalty := penalty }
        tac := res.tac }
  | simpEntries es =>
    return #[RuleSetMember'.normSimpEntries es]


structure SafeRuleClauses where
  builder : RegularBuilderClause
  deriving Inhabited

namespace SafeRuleClauses

def ofClauses (cs : Clauses) : m SafeRuleClauses := do
  let builder ←
    match cs.builder with
    | some (BuilderClause.regular b) => pure b
    | some b => errInvalidBuilder b
    | none => pure RegularBuilderClause.safeDefault
  return { builder := builder }
  where
    errInvalidBuilder {α} (b : BuilderClause) : m α :=
      throwError "aesop: {b.builderName} builder cannot be used with safe rules"

end SafeRuleClauses

def buildSafeRule (penalty : Option Int) (clauses : SafeRuleClauses)
    (i : RuleIdent) : MetaM (Array RuleSetMember) := do
  let results ← clauses.builder.toRuleBuilder i
  let penalty := penalty.getD 0
  results.mapM λ res => RuleSetMember'.safeRule
    { name := `safe ++ res.builderName ++ i.ruleName
      indexingMode := res.indexingMode,
      usesBranchState := res.mayUseBranchState
      extra := { penalty := penalty, safety := Safety.safe }
        -- TODO support almost_safe rules
      tac := res.tac }


structure UnsafeRuleClauses where
  builder : RegularBuilderClause
  deriving Inhabited

namespace UnsafeRuleClauses

def ofClauses (cs : Clauses) : m UnsafeRuleClauses := do
  let builder ←
    match cs.builder with
    | some (BuilderClause.regular b) => pure b
    | some b => errInvalidBuilder b
    | none => pure RegularBuilderClause.unsafeDefault
  return { builder := builder }
  where
    errInvalidBuilder {α} (b : BuilderClause) : m α :=
      throwError "aesop: {b.builderName} builder cannot be used with unsafe rules"

end UnsafeRuleClauses

def buildUnsafeRule (successProbability : Percent) (clauses : UnsafeRuleClauses)
    (i : RuleIdent) : MetaM (Array RuleSetMember) := do
  let results ← clauses.builder.toRuleBuilder i
  results.mapM λ res => RuleSetMember'.unsafeRule
    { name := `unsafe ++ res.builderName ++ i.ruleName
      indexingMode := res.indexingMode,
      usesBranchState := res.mayUseBranchState
      extra := { successProbability := successProbability },
      tac := res.tac }


inductive RuleConfig
  | norm (penalty : Int) (clauses : NormRuleClauses)
  | safe (penalty : Int) (clauses : SafeRuleClauses)
  | «unsafe» (successProbability : Percent) (clauses : UnsafeRuleClauses)
  deriving Inhabited

namespace RuleConfig

protected def ofKindAndClauses (kind : RuleKind) (clauses : Array Clause) :
    m RuleConfig := do
  let clauses := Clauses.ofClauseArray clauses
  match kind with
  | RuleKind.norm penalty =>
    norm penalty <$> NormRuleClauses.ofClauses clauses
  | RuleKind.safe penalty =>
    safe penalty <$> SafeRuleClauses.ofClauses clauses
  | RuleKind.unsafe successProbability =>
    «unsafe» successProbability <$> UnsafeRuleClauses.ofClauses clauses

protected def parse : Syntax → m RuleConfig
  | `(attr|aesop $kind:aesop_kind $[$clauses:aesop_clause]*) => do
    let kind ← RuleKind.parse kind
    let clauses ← clauses.mapM Clause.parse
    RuleConfig.ofKindAndClauses kind clauses
  | _ => unreachable!

protected def applyToRuleIdent (i : RuleIdent) : RuleConfig →
    MetaM (Array RuleSetMember)
  | norm penalty clauses => buildNormRule penalty clauses i
  | safe penalty clauses => buildSafeRule penalty clauses i
  | «unsafe» successProbability clauses =>
    buildUnsafeRule successProbability clauses i

end RuleConfig

structure AesopAttr where
  ext : ScopedEnvExtension (RuleSetMember' RuleTacDescr) RuleSetMember RuleSet
  impl : AttributeImpl
  deriving Inhabited

initialize attr : AesopAttr ← do
  let ext ← registerScopedEnvExtension {
    name := `aesop
    mkInitial := return {}
    ofOLeanEntry := λ rs r => runMetaMAsImportM r.ofDescr
    toOLeanEntry := λ r => r.toDescr.getD
      (panic! "aesop attribute extension: trying to serialise a rule set member without a description")
    addEntry := λ rs r => rs.add r
  }
  let impl : AttributeImpl := {
    name := `aesop
    descr := "Register a declaration as an Aesop rule."
    add := λ decl stx attrKind => do
      let config ← RuleConfig.parse stx
      let rules ← runMetaMAsCoreM $ config.applyToRuleIdent (RuleIdent.const decl)
      rules.forM λ rule => ext.add rule attrKind
    erase := λ _ =>
      throwError "aesop attribute currently cannot be removed"
  }
  -- Despite the name, this also works for non-builtin attributes.
  registerBuiltinAttribute impl
  pure { ext := ext, impl := impl }

def getAttrRuleSet : CoreM RuleSet := do
  attr.ext.getState (← getEnv)

def getRuleSet : MetaM RuleSet := do
  let defaultSimpLemmas ← getSimpLemmas
  let rs ← getAttrRuleSet
  return { rs with normSimpLemmas := defaultSimpLemmas.merge rs.normSimpLemmas }


namespace Parser.Tactic

declare_syntax_cat aesop_rule

syntax ident (aesop_prio)? (aesop_clause)* : aesop_rule

declare_syntax_cat aesop_tactic_clause

syntax ruleList := "[" aesop_rule,+,? "]"

syntax "(" &"unsafe" ruleList ")" : aesop_tactic_clause
syntax "(" &"safe"   ruleList ")" : aesop_tactic_clause
syntax "(" &"norm"   ruleList ")" : aesop_tactic_clause
syntax "(" &"options" term ")" : aesop_tactic_clause

syntax (name := aesop) &"aesop " (aesop_tactic_clause)* : tactic

end Parser.Tactic

structure AdditionalRule where
  ruleIdent : RuleIdent
  config : RuleConfig
  deriving Inhabited

namespace AdditionalRule

protected def parse (prioParser : Option Syntax → Except String α)
    (ruleKind : α → RuleKind) : Syntax → MetaM AdditionalRule
  | `(aesop_rule|$i:ident $[$prio:aesop_prio]? $clauses:aesop_clause*) => do
    let prio ←
      match prioParser prio with
      | Except.ok p => p
      | Except.error e => throwError "aesop: at rule {i}: {e}"
    let clauses ← clauses.mapM Clause.parse
    let config ← RuleConfig.ofKindAndClauses (ruleKind prio) clauses
    return { ruleIdent := (← RuleIdent.ofName i.getId), config := config }
  | _ => unreachable!

protected def toRuleSetMember (r : AdditionalRule) :
    MetaM (Array RuleSetMember) :=
  r.config.applyToRuleIdent r.ruleIdent

end AdditionalRule

inductive TacticClause
  | additionalRules (rs : Array AdditionalRule)
  | options (opts : Aesop.Options)

namespace TacticClause

def parse : Syntax → TermElabM TacticClause
  | `(aesop_tactic_clause|(unsafe [$rules:aesop_rule,*])) => additionalRules <$>
    (rules : Array Syntax).mapM λ stx =>
      AdditionalRule.parse parsePrioForUnsafeRule RuleKind.unsafe stx
  | `(aesop_tactic_clause|(safe [$rules:aesop_rule,*])) => additionalRules <$>
    (rules : Array Syntax).mapM λ stx =>
      AdditionalRule.parse parsePrioForSafeRule RuleKind.safe stx
  | `(aesop_tactic_clause|(norm [$rules:aesop_rule,*])) => additionalRules <$>
    (rules : Array Syntax).mapM λ stx =>
      AdditionalRule.parse parsePrioForNormRule RuleKind.norm stx
  | `(aesop_tactic_clause|(options $t:term)) => do
    let e ← elabTerm t (some (mkConst ``Aesop.Options))
    return options (← evalOptionsExpr e)
  | _ => unreachable!

end TacticClause


structure TacticConfig where
  additionalRules : Array AdditionalRule := #[]
  options : Aesop.Options := {}
  deriving Inhabited

namespace TacticConfig

def addTacticClause (conf : TacticConfig) : TacticClause → TacticConfig
  | TacticClause.additionalRules rs =>
    { conf with additionalRules := conf.additionalRules ++ rs }
  | TacticClause.options opts =>
    { conf with options := opts }

def ofTacticClauses (clauses : Array TacticClause) : TacticConfig :=
  clauses.foldl (init := {}) addTacticClause

-- NOTE: Must be called in the MVar context of the main goal.
protected def parse : Syntax → TermElabM TacticConfig
  | `(tactic|aesop $[$clauses:aesop_tactic_clause]*) =>
    return ofTacticClauses (← clauses.mapM TacticClause.parse)
  | _ => unreachable!

def additionalRuleSetMembers (c : TacticConfig) : MetaM (Array RuleSetMember) :=
  c.additionalRules.concatMapM (·.toRuleSetMember)

end TacticConfig

end Aesop
