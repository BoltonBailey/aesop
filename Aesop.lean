import Aesop.BestFirstSearch
import Aesop.Config
import Aesop.DefaultRules
import Aesop.Main
import Aesop.MutAltTree
import Aesop.Percent
import Aesop.RuleBuilder
import Aesop.Rule
import Aesop.RuleTac
import Aesop.Tracing
import Aesop.Tree
import Aesop.Util
