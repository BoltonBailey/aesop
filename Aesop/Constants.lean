/-
Copyright <redacted>
Released under Apache 2.0 license as described in the file LICENSE.
Authors: <redacted>
-/

import Aesop.Percent

namespace Aesop

def unificationGoalPenalty : Percent :=
  ⟨0.8⟩

def postponedSafeRuleSuccessProbability : Percent :=
  ⟨0.9⟩

end Aesop
