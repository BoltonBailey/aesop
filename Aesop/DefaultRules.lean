/-
Copyright (c) 2021 Jannis Limperg. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Jannis Limperg
-/

import Aesop.DefaultRules.Assumption
import Aesop.DefaultRules.ApplyHyps
import Aesop.DefaultRules.SplitHyps
import Aesop.Config

open Lean
open Lean.Elab
open Lean.Elab.Tactic

namespace Aesop.DefaultRules

-- TODO avoid TacticM?
def intros : TacticM Unit := do
  evalTactic (← `(tactic|intros))

def splitHyps : UserRuleTac := λ input =>
  return { regularGoals := #[(← splitAllHyps input.goal).snd] }

end DefaultRules

-- TODO As soon as the Aesop rule supports named rule sets, we can just
-- tag the above tactics with `@[aesop ... (rule_set default)]` or something.
def defaultRules : TermElabM (Array RuleSetMember) := do
  mkRules #[
    (``DefaultRules.assumption, ← `(attr|aesop safe -50)),
      -- The assumption rule is subsumed by applyHyps. But we still want to have
      -- it separately as a low-penalty safe rule so that very easy goals are
      -- discharged immediately.
    (``DefaultRules.intros    , ← `(attr|aesop norm -1)),
    (``DefaultRules.splitHyps , ← `(attr|aesop norm 0)),
    (``DefaultRules.applyHyps , ← `(attr|aesop 75%))]
  where
    mkRule (decl : Name) (configStx : Syntax) : TermElabM RuleSetMember := do
      let config ← RuleConfig.parse configStx
      config.applyToRuleIdent (RuleIdent.const decl)

    mkRules (rs : Array (Name × Syntax)) : TermElabM (Array RuleSetMember) :=
      rs.mapM (λ (decl, configStx) => mkRule decl configStx)

end Aesop
