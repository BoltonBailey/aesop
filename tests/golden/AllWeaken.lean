/-
Copyright <redacted>
Released under Apache 2.0 license as described in the file LICENSE.
Authors: <redacted>
-/

import Aesop

set_option aesop.check.all true

inductive All (P : α → Prop) : List α → Prop where
  | none : All P []
  | more {x xs} : P x → All P xs → All P (x :: xs)

@[aesop unsafe]
axiom weaken {α} (P Q : α → Prop) (wk : ∀ x, P x → Q x) (xs : List α)
  (h : All P xs) : All Q xs

example : All (· ∈ []) (@List.nil α) := by
  aesop (options := { maxRuleApplications := 50, terminal := true })
