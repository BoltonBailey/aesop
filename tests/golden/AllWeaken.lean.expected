tests/golden/AllWeaken.lean:20:2: error: aesop: maximum number of rule applications (50) reached. Set the 'maxRuleApplications' option to increase the limit.
